src/main/java/co/cask/gcp/bigquery/BigQuerySink.java

```diff
@@ -442,4 +451,104 @@ private void decodeSimpleTypes(JsonObject json, String name, StructuredRecord in
+
+    TableId id = TableId.of(project, config.dataset, config.table);
+    Table table = bigquery.getTable(id);
+    if (table == null) {
+      // Table does not exist, so no further validation is required.
+      return;
+    }
+
+    com.google.cloud.bigquery.Schema bqSchema = table.getDefinition().getSchema();
+    if (bqSchema == null) {
+      // Table is created without schema, so no further validation is required.
+      return;
+    }
+
+    FieldList bqFields = bqSchema.getFields();
+    List<Schema.Field> outputSchemaFields = Schema.parseJson(config.schema).getFields();
+
+    if (bqFields.size() != outputSchemaFields.size()) {
+      throw new IllegalArgumentException(
+        "Output schema does not match BigQuery table schema. Output schema must be compatible with " +
+          " BigQuery table schema.");
+    }
+
+    List<String> diff = new ArrayList<>();
+
+    for (Schema.Field field : outputSchemaFields) {
+      diff.add(field.getName());
+    }
+
+    for (Field field : bqFields) {
+      diff.remove(field.getName());
+    }
+
+    // Output schema should not have fields that are not present in BigQuery table.
+    if (!diff.isEmpty()) {
+      throw new IllegalArgumentException(
+        String.format("Output schema has more fields %s than BigQuery table %s schema. " +
+                        "Output schema must be compatible with BigQuery table schema.", diff, config.table));
+    }
+
+    Map<Schema.Type, LegacySQLTypeName> typeMap = new HashMap<>();
+    typeMap.put(Schema.Type.INT, LegacySQLTypeName.INTEGER);
+    typeMap.put(Schema.Type.LONG, LegacySQLTypeName.INTEGER);
+    typeMap.put(Schema.Type.STRING, LegacySQLTypeName.STRING);
+    typeMap.put(Schema.Type.FLOAT, LegacySQLTypeName.FLOAT);
+    typeMap.put(Schema.Type.DOUBLE, LegacySQLTypeName.FLOAT);
+    typeMap.put(Schema.Type.BOOLEAN, LegacySQLTypeName.BOOLEAN);
+    typeMap.put(Schema.Type.BYTES, LegacySQLTypeName.BYTES);
+
+    Map<Schema.LogicalType, LegacySQLTypeName> logicalTypeMap = new HashMap<>();
+    logicalTypeMap.put(Schema.LogicalType.DATE, LegacySQLTypeName.DATE);
+    logicalTypeMap.put(Schema.LogicalType.TIME_MILLIS, LegacySQLTypeName.TIME);
+    logicalTypeMap.put(Schema.LogicalType.TIME_MICROS, LegacySQLTypeName.TIME);
+    logicalTypeMap.put(Schema.LogicalType.TIMESTAMP_MILLIS, LegacySQLTypeName.TIMESTAMP);
+    logicalTypeMap.put(Schema.LogicalType.TIMESTAMP_MICROS, LegacySQLTypeName.TIMESTAMP);
+
+    // validate type of fields against bigquery column types
+    for (Schema.Field field : outputSchemaFields) {
+      Schema fieldSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable() : field.getSchema();
+      Schema.Type type = fieldSchema.getType();
+      Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+
+      if (logicalType != null) {
+        if (logicalTypeMap.get(logicalType) == null) {
+          throw new IllegalArgumentException(String.format("Type %s is not supported", logicalType));
```

Comment:
include the field name in the error too.

Field 'abc' is of unsupported type 'logical type'